pydis_site/apps/main/views/home.py

```diff
@@ -0,0 +1,101 @@
+                    "forks_count": repo["forks_count"],
+                    "stargazers_count": repo["stargazers_count"],
+                }
+        return repo_dict
+
+    def _get_repo_data(self):
+        """Build a list of RepositoryMetadata objects that we can use to populate the front page."""
+
+        # Try to get site data from the cache
+        try:
+            repo_data = RepositoryMetadata.objects.get(repo_name="python-discord/site")
+
+            # If the data is older than 2 minutes, we should refresh it.
+            if (timezone.now() - repo_data.last_updated).seconds > 120:
+
+                # Get new data from API
+                api_data_container = self._get_api_data()
+                repo_data_container = []
+
+                # Update or create all RepoData objects in settings.HOMEPAGE_REPOS
+                for repo_name, api_data in api_data_container.items():
+                    try:
+                        repo_data = RepositoryMetadata.objects.get(repo_name=repo_name)
+                        repo_data.description = api_data["description"]
+                        repo_data.language = api_data["language"]
+                        repo_data.forks = api_data["forks_count"]
+                        repo_data.stargazers = api_data["stargazers_count"]
+                    except RepositoryMetadata.DoesNotExist:
+                        repo_data = RepositoryMetadata(
+                            repo_name=api_data["full_name"],
+                            description=api_data["description"],
+                            forks=api_data["forks_count"],
+                            stargazers=api_data["stargazers_count"],
+                            language=api_data["language"],
+                        )
+                    repo_data.save()
+                    repo_data_container.append(repo_data)
+                return repo_data_container
+
+            # Otherwise, if the data is fresher than 2 minutes old, we should just return it.
+            else:
+                return RepositoryMetadata.objects.all()
+
+        # If this is raised, the database has no repodata at all, we will create them all.
+        except RepositoryMetadata.DoesNotExist:
+
+            # Get new data from API
+            api_data_container = self._get_api_data()
+            repo_data_container = []
+
+            # Create all the repodata records in the database.
+            for api_data in api_data_container.values():
+                repo_data = RepositoryMetadata(
+                    repo_name=api_data["full_name"],
+                    description=api_data["description"],
+                    forks=api_data["forks_count"],
+                    stargazers=api_data["stargazers_count"],
+                    language=api_data["language"],
+                )
+                repo_data.save()
+                repo_data_container.append(repo_data)
+
+            return repo_data_container
```

Comment:
this reads like lots of duplicated code, maybe `get_or_create` could help?

Reply:
get_or_create can't solve this problem, because the entire point is that I only want to call the API if the get fails. with get_or_create I would need to call the API anyway just so I have something to put into the create condition in case get fails. You'll have to PR in a DRYer solution yourself if you want to improve it because I don't have one, and I already spent enough time looking for one.